// @ciscospark/react-redux-spark/src/inject-spark.js
import React, {Component} from 'react';
import PropTypes from 'prop-types';

import SparkComponent from './component';

function getDisplayName(C) {
  return C.displayName || C.name || `C`;
}

export default function injectSpark(WrappedComponent, options = {}) {
  const {
    withRef = false
  } = options;

  class InjectSpark extends Component {

    shouldComponentUpdate(nextProps) {
      return nextProps !== this.props;
    }

    getWrappedInstance() {
      // eslint-disable-next-line react/no-string-refs
      return this.refs.wrappedInstance;
    }

    render() {
      const {accessToken, account} = this.props;

      return (
        <div>
          {
            // check here for account id or accessToken
            (accessToken || account) &&
            <SparkComponent accessToken={accessToken} account={account} {...this.props} />
          }
          <WrappedComponent
            {...this.props}
            ref={withRef ? `wrappedInstance` : null}
          />
        </div>
      );
    }
  }

  // should take refresh token, interaction, openInteraction etc...
  InjectSpark.propTypes = {
    accessToken: PropTypes.string,
    account: PropTypes.string
  };

  InjectSpark.displayName = `InjectSpark(${getDisplayName(WrappedComponent)})`;
  InjectSpark.WrappedComponent = WrappedComponent;

  return InjectSpark;
}
