// @altocloud/react-redux-guest/src/actions.js
import {
  fetchFreeUser,
  setUserAsBusy
} from './guest';

export const GET_GUEST_USER_BEGIN = `GET_GUEST_USER_BEGIN`;
export function getGuestUserBegin() {
  return {
    type: GET_GUEST_USER_BEGIN,
    payload: {
      isFetchingGuest: true
    }
  };
}

export const GET_GUEST_USER = `GET_GUEST_USER`;
export function getGuestUser(guest) {
  return {
    type: GET_GUEST_USER,
    payload: {
      isFetchingGuest: false,
      guest
    }
  };
}

export const UPDATE_GUEST_USER = `UPDATE_GUEST_USER`;
export function updateGuestUser(patch) {
  return {
    type: UPDATE_GUEST_USER,
    payload: {
      patch
    }
  };
}

export function fetchGuestUser(account) {
  return (dispatch) => {
    dispatch(getGuestUserBegin());
    fetchFreeUser(account).then((guest) => {
      dispatch(getGuestUser(guest));
    });
  };
}

export function updateGuestUserInteraction(account, sparkId, interaction) {
  return (dispatch) => {
    setUserAsBusy(account, sparkId, interaction).then(() => {
      dispatch(updateGuestUser({interaction}));
    });
  };
}
