// @altocloud/widget-customer-video/src/root.js
import React, {PropTypes} from 'react';
import {Provider} from 'react-redux';
import MessageMeetWidgetIntl from './intl-container';
import store from './store';

export default function Root({
  accessToken,
  closeInteraction,
  refreshToken,
  initialActivity,
  interaction,
  onEvent,
  openInteraction,
  startCall,
  toPersonEmail,
  toPersonId,
  updateInteraction
}) {
  console.log(`WidgetCustomerVideo Root: `, toPersonEmail);
  return (
    <Provider store={store}>
      <MessageMeetWidgetIntl
        accessToken={accessToken}
        closeInteraction={closeInteraction}
        initialActivity={initialActivity}
        interaction={interaction}
        onEvent={onEvent}
        openInteraction={openInteraction}
        refreshToken={refreshToken}
        startCall={startCall}
        toPersonEmail={toPersonEmail}
        toPersonId={toPersonId}
        updateInteraction={updateInteraction}
      />
    </Provider>
  );
}

Root.propTypes = {
  accessToken: PropTypes.string.isRequired,
  closeInteraction: PropTypes.func.isRequired,
  initialActivity: PropTypes.string,
  interaction: PropTypes.object.isRequired,
  onEvent: PropTypes.func,
  openInteraction: PropTypes.func.isRequired,
  refreshToken: PropTypes.string.isRequired,
  startCall: PropTypes.bool,
  toPersonEmail: PropTypes.string,
  toPersonId: PropTypes.string,
  updateInteraction: PropTypes.func.isRequired
};
