// @altocloud/alto-spark-video/src/container.js
/* eslint-disable react/no-set-state */
import React, {Component, PropTypes} from 'react';
import {connect} from 'react-redux';
import {IntlProvider} from 'react-intl';
import messages from './locales/en';

import classNames from 'classnames';
import WidgetCustomerVideo from '@altocloud/widget-customer-video';
import LoadingScreen from '@ciscospark/react-component-loading-screen';

import {injectGuest} from '@altocloud/react-redux-guest';

import styles from './styles.css';

class AltoSparkVideo extends Component {
  constructor(props) {
    super(props);
    const l = window.location;
    const redirectUri = `${l.protocol}//${l.host}${l.pathname}`.replace(/\/$/, ``);
    const clientId = process.env.CISCOSPARK_CLIENT_ID;
    const clientSecret = process.env.CISCOSPARK_CLIENT_SECRET;
    this.state = {
      accessToken: props.accessToken || null,
      refreshToken: props.refreshToken || null,
      authenticate: false,
      running: false,
      clientId,
      clientSecret,
      scope: `spark:kms spark:rooms_read spark:rooms_write spark:memberships_read spark:memberships_write spark:messages_read spark:messages_write`,
      redirectUri,
      error: ``
    };
    // no need to fetch free user if access token provided
    this.hasSparkAccountToken = !!props.accessToken;
  }

  componentWillReceiveProps(nextProps) {
    if (nextProps.guest.get(`guest`)) {
      const guest = nextProps.guest.get(`guest`).toJS();
      this.setState({...guest});
    }
  }

  shouldComponentUpdate() {
    return true;
  }

  getLoadingMessage() {
    if (this.state.error) {
      return this.state.error;
    }
    else if (!this.props.interaction) {
      return `Waiting for Interaction Details`;
    }
    else if (this.props.interaction.state === `closed`) {
      return `Interaction has been closed by the agent, please start a new interaction`;
    }
    else if (!this.state.accessToken) {
      return `Authenticating`;
    }
    return `Starting Cisco Spark`;
  }

  render() {
    const loadingMessage = this.getLoadingMessage();

    if (this.state.accessToken) {
      return (
        <IntlProvider locale={`en`} messages={messages}>
          <div className={classNames(`widget-component-container`, styles.widgetComponentContainer)}>
            <WidgetCustomerVideo
              accessToken={this.state.accessToken}
              closeInteraction={this.props.closeInteraction}
              interaction={this.props.interaction}
              openInteraction={this.props.openInteraction}
              refreshToken={this.state.refreshToken}
              toPersonEmail={this.props.toPersonEmail}
              updateInteraction={this.props.updateInteraction}
            />
          </div>
        </IntlProvider>);
    }
    return (
      <div className={classNames(`widget-component-container`, styles.widgetComponentContainer)}>
        <LoadingScreen loadingMessage={loadingMessage} />
      </div>);
  }
}

const injectedPropTypes = {
  guest: PropTypes.object.isRequired
};

AltoSparkVideo.propTypes = {
  accessToken: PropTypes.string,
  closeInteraction: PropTypes.func,
  interaction: PropTypes.object,
  openInteraction: PropTypes.func,
  refreshToken: PropTypes.string,
  toPersonEmail: PropTypes.string,
  updateInteraction: PropTypes.func,
  ...injectedPropTypes
};

AltoSparkVideo.title = `Altocloud Customer Video`;
AltoSparkVideo.path = `/wmm-demo`;

function mapStateToProps(state) {
  return {
    guest: state.guest,
    guestStatus: state.guest.get(`status`).toJS()
  };
}

export default connect(mapStateToProps, () => ({}))(injectGuest(AltoSparkVideo));
