// @altocloud/alto-spark-video/src/container.js
/* eslint-disable react/no-set-state */
import React, {Component, PropTypes} from 'react';
import {IntlProvider} from 'react-intl';
import messages from './locales/en';

import classNames from 'classnames';
import SparkGuest from '@altocloud/spark-guest';
import WidgetCustomerVideo from '@altocloud/widget-customer-video';
import LoadingScreen from '@ciscospark/react-component-loading-screen';

import styles from './styles.css';

class AltoSparkVideoComponent extends Component {
  constructor(props) {
    super(props);
    const l = window.location;
    const redirectUri = `${l.protocol}//${l.host}${l.pathname}`.replace(/\/$/, ``);
    const clientId = process.env.CISCOSPARK_CLIENT_ID;
    const clientSecret = process.env.CISCOSPARK_CLIENT_SECRET;
    this.state = {
      accessToken: props.accessToken || null,
      authenticate: false,
      running: false,
      clientId,
      clientSecret,
      scope: `spark:kms spark:rooms_read spark:rooms_write spark:memberships_read spark:memberships_write spark:messages_read spark:messages_write`,
      redirectUri
    };
    // no need to fetch free user if access token provided
    this.hasSparkAccountToken = !!props.accessToken;
  }

  shouldComponentUpdate() {
    return true;
  }

  getLoadingMessage() {
    if (!this.props.account) {
      return `Getting Account Details`;
    }
    else if (!this.props.interaction) {
      return `Getting Interaction Details`;
    }
    else if (!this.state.accessToken) {
      return `Authenticating`;
    }
    return `Loading`;
  }

  getFreeGuestAccount(account, interaction) {
    // already authenticated, do nothing
    if (this.state.accessToken) {
      return;
    }
    // no interaction or account, do nothing
    if (!interaction || !interaction.id || !account) {
      return;
    }
    // get free user
    this.hasSparkAccountToken = true;
    const sparkGuest = new SparkGuest(account.id);
    sparkGuest.getFreeUser().then(({sparkId, accessToken, refreshToken}) => {
      sparkGuest.setUserAsBusy(account.id, sparkId, interaction).then(() => {
        this.setState({
          accessToken,
          refreshToken
        });
      });
    });
  }

  render() {
    if (!this.hasSparkAccountToken) {
      this.getFreeGuestAccount(this.props.account, this.props.interaction);
    }

    const loadingMessage = this.getLoadingMessage();

    if (this.state.accessToken) {
      return (
        <IntlProvider locale={`en`} messages={messages}>
          <div className={classNames(`widget-component-container`, styles.widgetComponentContainer)}>
            <WidgetCustomerVideo
              accessToken={this.state.accessToken}
              closeInteraction={this.props.closeInteraction}
              interaction={this.props.interaction}
              openInteraction={this.props.openInteraction}
              refreshToken={this.state.refreshToken}
              toPersonEmail={this.props.toPersonEmail}
              updateInteraction={this.props.updateInteraction}
            />
          </div>
        </IntlProvider>);
    }
    return (
      <div className={classNames(`widget-component-container`, styles.widgetComponentContainer)}>
        <LoadingScreen loadingMessage={loadingMessage} />
      </div>);
  }
}

AltoSparkVideoComponent.propTypes = {
  accessToken: PropTypes.string,
  account: PropTypes.object,
  closeInteraction: PropTypes.func.isRequired,
  interaction: PropTypes.object,
  openInteraction: PropTypes.func.isRequired,
  toPersonEmail: PropTypes.string,
  updateInteraction: PropTypes.func.isRequired
};

AltoSparkVideoComponent.title = `Altocloud Customer Video`;
AltoSparkVideoComponent.path = `/wmm-demo`;

export default AltoSparkVideoComponent;
