// @altocloud/alto-spark-video/src/index.js
/* eslint-disable react/no-set-state */
import React, {Component} from 'react';

import {IntlProvider} from 'react-intl';
import messages from './locales/en';

import classNames from 'classnames';
import SparkGuest from '@altocloud/spark-guest';
import WidgetCustomerVideo from '@altocloud/widget-customer-video';
import LoadingScreen from '@ciscospark/react-component-loading-screen';

import styles from './styles.css';

class AltoSparkVideo extends Component {
  constructor(props) {
    super(props);
    const l = window.location;
    const redirectUri = `${l.protocol}//${l.host}${l.pathname}`.replace(/\/$/, ``);
    const clientId = process.env.CISCOSPARK_CLIENT_ID;
    const clientSecret = process.env.CISCOSPARK_CLIENT_SECRET;
    this.state = {
      authenticate: false,
      running: false,
      clientId,
      clientSecret,
      scope: `spark:kms spark:rooms_read spark:rooms_write spark:memberships_read spark:memberships_write spark:messages_read spark:messages_write`,
      redirectUri
    };
    this.sparkGuest = new SparkGuest(`9c42eeeb`);
  }

  componentWillMount() {
    this.sparkGuest.getFreeUser().then((data) => {
      this.setState({
        accessToken: data.accessToken,
        refreshToken: data.refreshToken
      });
    });
    // this.setState({
    //   accessToken: `YTE2MjYwOTUtZWY4MC00ODBiLWEyNWQtNzYzMjcwMjJhZjMwNjdjMzQwZDYtNzBk`,
    //   refreshToken: `ODllYjdiMDYtMWM3NS00NGE1LTg5YmEtOTM4N2U5MDIyNDM2Yzg1ZDgzMzAtN2E0`
    // });
  }

  shouldComponentUpdate() {
    return true;
  }

  render() {
    if (this.state.accessToken && this.state.refreshToken) {
      return (
        <IntlProvider locale={`en`} messages={messages}>
          <div className={classNames(`widget-component-container`, styles.widgetComponentContainer)}>
            <WidgetCustomerVideo accessToken={this.state.accessToken} refreshToken={this.state.refreshToken} />
          </div>
        </IntlProvider>);
    }
    return (
      <div className={classNames(`widget-component-container`, styles.widgetComponentContainer)}>
        <LoadingScreen loadingMessage={`Getting OAuth Token`} />
      </div>);
  }
}

AltoSparkVideo.title = `Altocloud Customer Video`;
AltoSparkVideo.path = `/wmm-demo`;

export default AltoSparkVideo;
