// @altocloud/alto-spark-video/src/root.js
import React, {PropTypes} from 'react';
import {Provider} from 'react-redux';
import AltoSparkVideoComponent from './container';
import store from './store';

export default function Root({
  accessToken,
  account,
  closeInteraction,
  interaction,
  openInteraction,
  refreshToken,
  toPersonEmail,
  updateInteraction
}) {
  return (
    <Provider store={store}>
      <AltoSparkVideoComponent
        accessToken={accessToken}
        account={account}
        closeInteraction={closeInteraction}
        interaction={interaction}
        openInteraction={openInteraction}
        refreshToken={refreshToken}
        toPersonEmail={toPersonEmail}
        updateInteraction={updateInteraction}
      />
    </Provider>
  );
}

Root.propTypes = {
  accessToken: PropTypes.string,
  account: PropTypes.object,
  closeInteraction: PropTypes.func,
  interaction: PropTypes.object,
  openInteraction: PropTypes.func,
  refreshToken: PropTypes.string,
  toPersonEmail: PropTypes.string,
  updateInteraction: PropTypes.func
};
